include/aspect/introspection.h

```diff
@@ -29,10 +29,265 @@
+               , ExcMessage("The last variable needs to have >0 blocks."));
+
+        n_components_ = component_index;
+        n_blocks_ = block_index;
+
+        fes.resize(variables_.size());
+        multiplicities.resize(variables_.size());
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            fes[i] = &*(variables_[i].fe);
+            multiplicities[i] = variables_[i].multiplicity;
+            variables_[i].component_mask=ComponentMask(n_components_, false);
+            for (unsigned int c=0; c<variables_[i].n_components(); ++c)
+              variables_[i].component_mask.set(c+variables_[i].first_component_index, true);
+          }
+
+        components_to_blocks_.clear();
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            for (unsigned int c=0; c<variables_[i].n_components(); ++c)
+              components_to_blocks_.push_back(variables_[i].block_index
+                                              + ((variables_[i].n_blocks>1)?c:0));
+          }
+      }
+
+      /**
+       * Return the variable with name @p name. Throws an exception if this variable
+       * does not exist.
+       */
+      const Variable<dim> &variable(const std::string &name) const
+      {
+        typename std::map<std::string, Variable<dim>*>::const_iterator it
+          = name_map.find(name);
+        Assert(it != name_map.end(), ExcMessage("Variable '" + name + "' not found!"));
+        return *(it->second);
+      }
+
+      /**
+       * Return the list of all variables.
+       */
+      const std::vector<Variable<dim> > &variables() const
+      {
+        return variables_;
+      }
+
+      /**
+       * Return the total number of components in the system.
+       */
+      unsigned int n_components() const
+      {
+        return n_components_;
+      }
+
+      /**
+       * Return the total number of block of the system.
+       */
+      unsigned int n_blocks() const
+      {
+        return n_blocks_;
+      }
+
+      /**
+       * Return the vector of finite element spaces used for the construction
+       * of the FESystem.
+       */
+      const std::vector<const FiniteElement<dim> *> &get_fes() const
+      {
+        return fes;
+      }
+
+      /**
+       * Return the vector of multiplicities used for the construction of the
+       * FESystem.
+       */
+      const std::vector<unsigned int> &get_multiplicities() const
+      {
+        return multiplicities;
+      }
+
+      /**
+       * Return a variable that describes for each vector component which vector
+       * block it corresponds to.
+       */
+      const std::vector<unsigned int> &components_to_blocks() const
+      {
+        return components_to_blocks_;
+      }
+
+    protected:
+      std::vector<Variable<dim> > variables_;
+
+      std::map<std::string, Variable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
```

Comment:
I'm still not a fan of `variable_`. In particular if only *some* variables use this convention.